### DISCUSSION TITLE
Getting trotterized exponential of a Hamiltonian

### DISCUSSION THREAD

**Discussion by einarG-1qbit** [goodchemistryco/Tangelo#227] [Q&A] [answered]
Hello,

1. I would like to load the molecule. I guess this is done with `SecondQuantizedMolecule`  
2. I would like to get the Hamiltonian for that molecule and get a circuit by exponentiation of the Hamiltonian and applying 1st order Trotterization. For that, I probably need `get_exponentiated_qubit_operator_circuit()` .
3. Next, I want to have full access to the generated circuit and be able to convert circuit gates into Clifford+T gate set or at least prepare the circuit for some conversion procedure that would be done outside Tangelo if that functionality is not present in the package.

As an exmaple, I use the molecule below.
```
H2 = [('H', (0, 0, 0)),('H', (0, 0, 0.74137727))]
mol_H2 = SecondQuantizedMolecule(H2, q=0, spin=0, basis="sto-3g")
```
The problem is that it is not obvious to me how to go from step 1 to step 2. Step 3 is also very non-obvious.

P.S. note that the resulting circuit will have multi-gates like $\exp (- i a t Y \otimes X \otimes X \otimes Y)$. Such gates are not supported in open qasm.

**Comment by alexfleury-sb** [comment, marked as answer]
Hi @einarG-1qbit! Thanks for your question: here is a more in-depth explanation (than yesterday's answer in slack):
1. Yes, you are right.
2. The first thing to do in that case is to convert the Hamiltonian from a fermion to a qubit basis. At the time of writing those lines, Tangelo supports the Jordan-Wigner (JW), Bravyi-Kitaev (BK), symmetry-conserving BK (scBK) and Jiang-Kalev-Mruczkiewicz-Neven (JKMN) mapping. The respective references should be included in the docstrings. 
This can be done with the `fermion_to_qubit_mapping` function (here an example where we want the JW mapping with the spin-up then spin-down orbital ordering).
```python
from tangelo.toolboxes.qubit_mappings.mapping_transform import fermion_to_qubit_mapping
...
qH = fermion_to_qubit_mapping(mol_H2.fermionic_hamiltonian, "JW", mol_H2.n_active_sos, mol_H2.n_active_electrons, up_then_down=True)
```
After, it is indeed the function `get_exponentiated_qubit_operator_circuit` that transforms the Hamiltonian into a circuit via trotterization.
```python
from tangelo.toolboxes.ansatz_generator.ansatz_utils import get_exponentiated_qubit_operator_circuit
exp_circ = get_exponentiated_qubit_operator_circuit(qH, time=1., trotter_order=1)
```
3. At this point, you have a `tangelo.linq.Circuit` object (basically it is a list of `tangelo.linq.Gate`). We have some helpers functions and methods to help manipulate (optimization, splitting, merging, etc.) the Circuit. However, the transpilation to a specific set of gates is not something explicitly supported for now. For this purpose, I have used [pytket](https://github.com/CQCL/pytket) in the past, and the circuits can be exported/imported with the `translator` functions for openqasm (https://github.com/goodchemistryco/Tangelo/blob/main/tangelo/linq/translator/translate_openqasm.py). There are also functions to convert from and to the qiskit, qulacs, cirq and other backend circuit formats. There is currently a [PR on that subject ](https://github.com/goodchemistryco/Tangelo/pull/223), and everything said above will be done from a single function called `translate_circuit`.

**Reply by einarG-1qbit** [reply]
Thanks Alexandre!
